File: BFS/Leetcode297.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 */
var serialize = function(root) {
    if(root == null) {
        return "{}";
    }
    const list = [];
    list.push(root);
    for(let i = 0 ; i < list.length; i++) {
        if(list[i] == null) {
            continue;
        }
        list.push(list[i].left);
        list.push(list[i].right);
    }
    let str = "{";
    str += list[0].val;
    for(let i = 1; i < list.length; i++) {
        str += ",";
        if(list[i] == null) {
            str += "#";
        } else {
            str += list[i].val;
        }
    }
    str += "}";
    return str;
};

/**
 * Decodes your encoded data to tree.
 *
 * @param {string} data
 * @return {TreeNode}
 */
var deserialize = function(data) {
    if(data === "{}") {
        return null;
    }
    // the type of arr's element is string
    const arr = data.slice(1, data.length - 1).split(",");
    const root = new TreeNode(Number.parseInt(arr[0]));
    let isLeft = true, index = 0;
    const list = [];
    list.push(root);
    for(let i = 1; i < arr.length; i++) {
        if(arr[i] !== "#") {
            const temp = new TreeNode(Number.parseInt(arr[i]));
            if(isLeft)
                list[index].left = temp;
            else
                list[index].right = temp;
            list.push(temp);
        }
        if(!isLeft) {
            index++;
        }
        isLeft = !isLeft;
    }
    return root;
};

/**
 * Your functions will be called as such:
 * deserialize(serialize(root));
 */

File: BFS/Leetcode297.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if(root == null) {
            return "{}";
        }
        List<TreeNode> list = new ArrayList<TreeNode>();
        list.add(root);
        for(int i = 0; i < list.size(); i++) {
            TreeNode temp = list.get(i);
            if(temp == null) {
                continue;
            }
            // when some nodes are null, they would be added into the list, but their children nodes wouldn't
            list.add(temp.left);
            list.add(temp.right);
        }
        // remove the null value at the tail
        while(list.get(list.size() - 1) == null) {
            list.remove(list.size() - 1);
        }
        StringBuilder str = new StringBuilder();
        str.append("{");
        str.append(list.get(0).val);
        // traverse the list again and get the final string
        for(int i = 1; i < list.size(); i++) {
            str.append(",");
            if(list.get(i) == null) {
                str.append("#");
            } else {
                str.append(list.get(i).val);
            }
        }
        str.append("}");
        return str.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if(data.equals("{}")) {
            return null;
        }
        List<TreeNode> list = new ArrayList<TreeNode>();
        String[] arr = data.substring(1, data.length()-1).split(",");
        TreeNode root = new TreeNode(Integer.parseInt(arr[0]));
        list.add(root);
        boolean isLeft = true;
        int index  = 0;
        for(int i = 1; i < arr.length; i++) {
            if(!arr[i].equals("#")){
                TreeNode temp = new TreeNode(Integer.parseInt(arr[i]));
                if(isLeft) {
                    list.get(index).left = temp;
                } else {
                    list.get(index).right = temp;
                }
                list.add(temp);
            }
            if(!isLeft) {
                index++;
            }
            isLeft = !isLeft;
        }
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));